File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/IntakeSubsystem.java
package org.firstinspires.ftc.teamcode.subsystems;

import com.acmerobotics.dashboard.config.Config;
import com.arcrobotics.ftclib.command.SubsystemBase;
import com.arcrobotics.ftclib.hardware.ServoEx;
import com.qualcomm.robotcore.hardware.CRServo;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.teamcode.util.LoggingUtils.FTCDashboardPackets;
import org.firstinspires.ftc.teamcode.util.RobotHardwareInitializer;

@Config
public class IntakeSubsystem extends SubsystemBase {

    DcMotorEx powerMotor;
    //ServoEx tilterServo;

    CRServo tilterServo;
    final double TILTER_POWER_SCALE = .5f;
    boolean currentState = false;
    final double MAX_POWER = 0.8;

    public static double INTAKE_POSITION_PADDING = 0; // "padding" from the max position of the servo

    public IntakeSubsystem(HardwareMap hardwareMap) throws Exception {
        this.powerMotor = RobotHardwareInitializer.MotorComponent.INTAKE_MOTOR.getEx(hardwareMap);
        tilterServo = RobotHardwareInitializer.CRServoComponent.INTAKE_TILTER.get(hardwareMap);
        tilterServo.setDirection(DcMotorSimple.Direction.FORWARD);
        //this.tilterServo = RobotHardwareInitializer.ServoComponent.INTAKE_TILTER.getEx(hardwareMap);
        //this.tilterServo.setInverted(true);
        //tilterServo.setRange(0, 35);
    }

    public void tiltIntake() {
        //tilterServo.setPosition(.7f);

    }

    public void untiltIntake() {
        //tilterServo.setPosition(0.4f);
    }

    public void moveTiltIntake(double power) {
        tilterServo.setPower(power * TILTER_POWER_SCALE);
    }

    public void setIntakeState(boolean activated, boolean reversed) {
        if (!reversed)
            powerMotor.setPower(activated ? MAX_POWER : 0);
        else
            powerMotor.setPower(activated ? -MAX_POWER : 0);
        currentState = activated;
    }

    public void toggleIntakeState() {
        setIntakeState(!currentState, false);
    }
    public void reverseIntake() {setIntakeState(!currentState, true);}
    public boolean getIntakeState() {
        return currentState;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/IntakeSubsystem.kt
package org.firstinspires.ftc.teamcode.subsystems

import com.acmerobotics.dashboard.config.Config
import com.arcrobotics.ftclib.command.SubsystemBase
import com.qualcomm.robotcore.hardware.CRServo
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.util.RobotHardwareInitializer

@Config
class IntakeSubsystem(hardwareMap: HardwareMap) : SubsystemBase() {
    var powerMotor: DcMotorEx =
        RobotHardwareInitializer.MotorComponent.INTAKE_MOTOR.getEx(hardwareMap)

    //ServoEx tilterServo;
    var tilterServo: CRServo = RobotHardwareInitializer.CRServoComponent.INTAKE_TILTER[hardwareMap]
    val TILTER_POWER_SCALE: Double = .5
    var intakeState: Boolean = false
    val MAX_POWER: Double = 0.8

    init {
        tilterServo.direction = DcMotorSimple.Direction.FORWARD
        //this.tilterServo = RobotHardwareInitializer.ServoComponent.INTAKE_TILTER.getEx(hardwareMap);
        //this.tilterServo.setInverted(true);
        //tilterServo.setRange(0, 35);
    }

    fun tiltIntake() {
        //tilterServo.setPosition(.7f);
    }

    fun untiltIntake() {
        //tilterServo.setPosition(0.4f);
    }

    fun moveTiltIntake(power: Double) {
        tilterServo.power = power * TILTER_POWER_SCALE
    }

    fun setIntakeState(activated: Boolean, reversed: Boolean) {
        if (!reversed) powerMotor.power = if (activated) MAX_POWER else 0.0
        else powerMotor.power = if (activated) -MAX_POWER else 0.0
        intakeState = activated
    }

    fun toggleIntakeState() {
        setIntakeState(!intakeState, false)
    }

    fun reverseIntake() {
        setIntakeState(!intakeState, true)
    }

    companion object {
        var INTAKE_POSITION_PADDING: Double = 0.0 // "padding" from the max position of the servo
    }
}
